Refactor blockpoller initialization to use first streamable block

billettc opened the pull request:
Replaced start block with first streamable block in BlockPoller to ensure accurate initiation. Added a check for the state file's existence and integrated state initialization with fetching the first streamable block. Updated tests accordingly.

maoueh commented on blockpoller/state_file.go:
```diff
@@ -33,20 +33,36 @@ type stateFile struct {
+	}
+	fp := filepath.Join(stateStorePath, "cursor.json")
+	_, err := os.Stat(fp)
+	exist := err == nil || os.IsExist(err)
```
I think `IsExists(err)` could be true only on doing a write, but `Stat` does a read only. It should be only `err == nil`

billettc committed:
Refactor BlockPoller to support generics and client handling

Refactor BlockPoller to support generic clients with type parameter 'C'. Updated associated methods, test functions, and interfaces to accommodate this change. This enhancement includes moving the retry mechanism and client handling to the poller, improving flexibility for different client types.

billettc committed:
Refactor block fetch functions to include context parameter

Updated several block fetch functions and utility methods to accept a context parameter for better cancellation and timeout control. Enhanced TestBlockClient to handle block fetching delays, and adjusted tests to accommodate these changes.

billettc committed:
Refactor Clients initialization to accept timeout duration

Updated the Clients struct to include a maxBlockFetchDuration parameter, enhancing flexibility in specifying timeout durations for block fetching. This change was reflected in the initialization function and associated test cases.

billettc committed:
Add rolling strategy support to RPC clients

Implemented new rolling strategy interfaces and updated RPC clients to utilize the rolling strategies. Introduced two strategies: RoundRobin and AlwaysUseFirst, modifying existing tests to incorporate these changes.

billettc committed:
Add MaxStopBlock variable to blockpoller

Introduce a MaxStopBlock variable to be used in the BlockPoller. This change helps manage the upper limit for block polling, allowing more flexibility and control over the polling process.

billettc committed:
Add sorting functionality and new rolling strategies

Introduced a sorting mechanism for clients in the `rpc` package. Added interfaces and implementations for sticky rolling strategy, along with comprehensive test cases. Improved thread safety and replaced the old rolling strategy implementations.

billettc committed:
Refactor sort function to use external sort value fetcher

Updated the Sort function to accept a SortValueFetcher interface and changed its implementations accordingly. This change improves flexibility and decouples sort value fetching logic from the client type. Also added zap logger dependency to NewClients initialization in tests.

billettc committed:
Define MaxStopBlock constant

Add a new MaxStopBlock constant set to math.MaxUint64. This change simplifies the code by using a predefined maximum stop block value for the poller logic.

billettc committed:
Remove unused MaxStopBlock constant

The MaxStopBlock constant was removed because it was not used anywhere in the code. Cleans up unnecessary variables and maintains code readability.

billettc committed:
Refactor testSortFetcher to use pointer receiver

billettc committed:
Rename fetchSortValue to FetchSortValue in SortValueFetcher.

This change makes the method public by capitalizing its name, allowing it to be accessed outside its package. This update is necessary for external packages that need to use this method.

billettc committed:
Reset rolling strategies state after sorting

Ensure the rolling strategies reset internal state variables such as `usedClientCount`, `nextClientIndex`, and others after encountering sorting errors. This is to guarantee proper client selection behavior in subsequent retries.

billettc committed:
Remove unnecessary locks and update test setups

billettc committed:
Update google-cloud-go dependency to latest version

Upgraded the `github.com/streamingfast/google-cloud-go` module to version `v0.0.0-20241202194114-f77ff78d4f66` in both `go.mod` and `go.sum`. This ensures the project is using the latest compatible changes from the dependency.
